src/librbd/io/ImageDispatcher.cc

```diff
@@ -191,13 +241,36 @@ bool ImageDispatcher<I>::send_dispatch(
     ImageDispatchSpec* image_dispatch_spec) {
   if (image_dispatch_spec->tid == 0) {
     image_dispatch_spec->tid = ++m_next_tid;
+
+    int r = util::clip_request(this->m_image_ctx, &image_dispatch_spec->image_extents);
+    if (r < 0) {
+      image_dispatch_spec->fail(r);
+      return true;
+    }
+  }
+
+  if ((image_dispatch_spec->dispatch_layer ==
+       IMAGE_DISPATCH_LAYER_WRITEBACK_CACHE) ||
+      (image_dispatch_spec->dispatch_layer == IMAGE_DISPATCH_LAYER_CORE)) {
+    bool finished = clip_request(image_dispatch_spec);
```

Comment:
Nit: why can't this be rolled into the above clip request? What's the purpose of (re)applying this as it comes through two layers? Can it just be applied once via the visitor?

Reply:
> Nit: why can't this be rolled into the above clip request?   

I can move it to above visitor.
> What's the purpose of (re)applying this as it comes through two layers? 

Please see the above comments https://github.com/ceph/ceph/pull/37014#issuecomment-696558022 
> Can it just be applied once via the visitor?

These IO processes will be either be handled in WRITEBACK layer or CORE layer.

Reply:
The core layer doesn't care that it's zero length -- it will get optimized away into a no-op via `set_request_count(0)`. So that leaves just the PWL layer that cares -- so perhaps it could just finish it early since it's the only one that cares.

Reply:
> The core layer doesn't care that it's zero-length -- it will get optimized away into a no-op via `set_request_count(0)`. 

Right. The core doesn't care whether it is zero-length, it handles it in AbstractImageWriteRequest<I>::send_request().  But maybe it is more efficient to check it here. 

> So that leaves just the PWL layer that cares -- \so perhaps it could just finish it early since it's the only one that cares.

In the updates, the preprocess is checked once. If with WRITEBACK layer, it sets the spec completed and finish. So the core layer will not check it again.   
```cpp
    if (total_bytes == 0) {
      image_dispatch_spec->dispatch_result = DISPATCH_RESULT_COMPLETE;
      image_dispatch_spec->aio_comp->set_request_count(0);
      return true;
    }
```
I am confused about the comments to finish it early.  Could you please give more explanation?

Reply:
Sorry, I mean if this is logic that is specific to PWL layer (i.e. it cannot properly handle no-op requests), perhaps we should move this logic down into the PWL dispatch layer (since the core dispatch layer doesn't care).

Reply:
Thank you for your explanation. I have moved the length check to PWL dispatch.